[Feature] Add H265 support with SDP, RTP payloader-deplayloader

niyatim23 opened the pull request:
*Issue #, if available:*
- Running the CI for https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1945 and addressing all the comments and fixes

*What was changed?*
- Added support for H265 in RTP and SDP

*Why was it changed?*
- This is new feature to expand support for different codecs

*How was it changed?*
- Introduced H265 payloader and de-payloader in RTP
- Modified the _SessionDescription.c_ to include H265 in offer and answer

*What testing was done for the changes?*
- Used #1963 with _kvsWebrtcClientMasterGstSample_ and _kvsWebrtcClientMaster_ as master and _kvsWebrtcClientViewerGstSample_ as viewer with _H265_ only and _H265/Opus_

```
Desktop % ffprobe -v error -show_entries stream=codec_name,sample_rate,width,height,channels -of default=noprint_wrappers=1 video.mkv
codec_name=hevc
width=1280
height=720
codec_name=opus
sample_rate=48000
channels=2
niyatim@6c7e67b7ca72 Desktop % 
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

niyatim23 commented on src/source/PeerConnection/SessionDescription.h:
```diff
@@ -45,11 +46,15 @@ extern "C" {
 
-#define DEFAULT_H264_FMTP   (PCHAR) "level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f"
+#define DEFAULT_H264_FMTP (PCHAR) "level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f"
+#define DEFAULT_H265_FMTP                                                                                                                            \
```
Need to discuss if we want to restrict the profile and other properties. Mostly inclined towards removing this but wanted everyone's opinion here

hassanctech replied:
webrtc is only mandated to support I think the constrained baseline profile, removing this would allow frames to be sent which are not part of webrtc spec.  I suspect we will need. specific FMTP line with properties/profile for H265 as well

hassanctech replied:
There is a draft RFC for H265 support in WebRTC:  https://datatracker.ietf.org/doc/draft-ietf-avtcore-hevc-webrtc/

It is only a draft but probably best to follow it for now, relevant section is:

```
   For the H.265 [H.265] codec, endpoints MUST support the payload
   formats defined in [RFC7798]. In addition, they MUST support Main
   Profile Level 3.1 (level-id=93) and SHOULD support Main Profile Level
   4 (level-id=120).
```

We need to make sure we have tested with these and we have space in our design to separate between different profiles as needed.

niyatim23 replied:
I tested this with the Android SDK and I don't receive an fmtp in the offer from libwebrtc:
```
2024-04-25 05:48:35.312 DEBUG   setRemoteDescription(): REMOTE_SDP:v=0
o=- 842804224343444729 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1 2
a=msid-semantic: WMS
m=audio 9 UDP/TLS/RTP/SAVPF 96 0 8 105 13 113 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=recvonly
a=rtcp-mux
a=rtpmap:96 AAC/16000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
m=video 9 UDP/TLS/RTP/SAVPF 127 97 98 99 100 101 102 103 104 125 106
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:12 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:11 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://tools.ietf.org/html/draft-ietf-avtext-framemarking-07
a=extmap:9 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=recvonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:127 VP8/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=127
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 H264/90000
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=fmtp:100 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:102 H265/90000
a=rtcp-fb:102 goog-remb
a=rtcp-fb:102 transport-cc
a=rtcp-fb:102 ccm fir
a=rtcp-fb:102 nack
a=rtcp-fb:102 nack pli
a=rtpmap:103 rtx/90000
a=fmtp:103 apt=102
a=rtpmap:104 red/90000
a=rtpmap:125 rtx/90000
a=fmtp:125 apt=104
a=rtpmap:106 ulpfec/90000
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag:wcqN
a=ice-pwd:v+PR8YjabtyDlJyDEMPSg3xW
a=ice-options:trickle renomination
a=fingerprint:sha-256 F3:4E:6E:A9:07:F4:C4:16:56:90:EE:5F:0C:C9:DD:80:64:3D:3B:59:24:7C:22:29:8C:FB:7D:BF:80:C4:26:4F
a=setup:actpass
a=mid:2
a=sctp-port:5000
a=max-message-size:262144

```
Should we be following the draft still?

niyatim23 replied:
Safari's offer with H265 doesn't have an fmtp at 104 (h265/90000) either:
```
[Log] PeerConnectionBackend::createOfferSucceeded(CCD61DEB87F60F75) Create offer succeeded:
v=0
o=- 985372420936776446 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1 2
a=extmap-allow-mixed
a=msid-semantic: WMS bc57d7ff-a695-4aa3-b860-7de2e4494fed
m=audio 9 UDP/TLS/RTP/SAVPF 111 63 9 0 8 13 110 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=sendrecv
a=msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 833dba34-201d-4710-b1ce-8273bf20a245
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:63 red/48000/2
a=fmtp:63 111/111
a=rtpmap:9 G722/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:126 telephone-event/8000
a=ssrc:2776556564 cname:VmWUiGQTNzJgTWxC
a=ssrc:2776556564 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 833dba34-201d-4710-b1ce-8273bf20a245
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 100 101 102 103 104 105 106 107 108 109 127 125 112 113 114
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:13 urn:3gpp:video-orientation
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:5 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:4 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:10 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:11 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=sendrecv
a=msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 H264/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=fmtp:96 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=640c1f
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:98 H264/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=fmtp:98 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 H264/90000
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=fmtp:100 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=640c1f
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:102 H264/90000
a=rtcp-fb:102 goog-remb
a=rtcp-fb:102 transport-cc
a=rtcp-fb:102 ccm fir
a=rtcp-fb:102 nack
a=rtcp-fb:102 nack pli
a=fmtp:102 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=42e01f
a=rtpmap:103 rtx/90000
a=fmtp:103 apt=102
a=rtpmap:104 H265/90000
a=rtcp-fb:104 goog-remb
a=rtcp-fb:104 transport-cc
a=rtcp-fb:104 ccm fir
a=rtcp-fb:104 nack
a=rtcp-fb:104 nack pli
a=rtpmap:105 rtx/90000
a=fmtp:105 apt=104
a=rtpmap:106 VP8/90000
a=rtcp-fb:106 goog-remb
a=rtcp-fb:106 transport-cc
a=rtcp-fb:106 ccm fir
a=rtcp-fb:106 nack
a=rtcp-fb:106 nack pli
a=rtpmap:107 rtx/90000
a=fmtp:107 apt=106
a=rtpmap:108 VP9/90000
a=rtcp-fb:108 goog-remb
a=rtcp-fb:108 transport-cc
a=rtcp-fb:108 ccm fir
a=rtcp-fb:108 nack
a=rtcp-fb:108 nack pli
a=fmtp:108 profile-id=0
a=rtpmap:109 rtx/90000
a=fmtp:109 apt=108
a=rtpmap:127 VP9/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=fmtp:127 profile-id=2
a=rtpmap:125 rtx/90000
a=fmtp:125 apt=127
a=rtpmap:112 red/90000
a=rtpmap:113 rtx/90000
a=fmtp:113 apt=112
a=rtpmap:114 ulpfec/90000
a=ssrc-group:FID 3773084415 3113708224
a=ssrc:3773084415 cname:VmWUiGQTNzJgTWxC
a=ssrc:3773084415 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
a=ssrc:3113708224 cname:VmWUiGQTNzJgTWxC
a=ssrc:3113708224 msid:bc57d7ff-a695-4aa3-b860-7de2e4494fed 094be066-b91c-4de0-adbe-b47bcf534826
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=ice-ufrag://pY
a=ice-pwd:svCwcEunmp3xe5js1YGHMWI8
a=ice-options:trickle
a=fingerprint:sha-256 64:61:0D:9D:DE:D0:37:5A:82:55:73:E9:41:06:30:79:C9:31:B0:5B:89:80:99:9D:B6:E9:61:F2:5D:1E:83:42
a=setup:actpass
a=mid:2
a=sctp-port:5000
a=max-message-size:262144
```

niyatim23 replied:
I verified the behavior for using fmtp. We send the default fmtp only if we are sending the offer ourselves first. If not, we emulate the behavior in the offer. If the offer doesn't contain fmtp for h265, we don't include it in the answer either. If it does, we include the same fmtp that we received in the answer. I think it should be fine to keep this in that case. However, we do not assume the fmtp specified here is our default profile and that is why the name is RTC_CODEC_H265 as opposed to H264's codec name

hassanctech replied:
Thanks for verifying the behavior.

hassanctech commented on .github/workflows/ci.yml:
```diff
@@ -612,7 +612,7 @@ jobs:
         run: |
           $env:Path += ';C:\webrtc\open-source\bin;C:\tools\pthreads-w32-2-9-1-release\Pre-built.2\dll\x64;C:\webrtc\build'
-          & "C:\webrtc\build\tst\webrtc_client_test.exe" --gtest_filter="-SignalingApiFunctionalityTest.receivingIceConfigOffer_SlowClockSkew:SignalingApiFunctionalityTest.iceServerConfigRefreshConnectedAuthExpiration:SignalingApiFunctionalityTest.receivingIceConfigOffer_FastClockSkew:SignalingApiFunctionalityTest.receivingIceConfigOffer_FastClockSkew_VerifyOffsetRemovedWhenClockFixed:DataChannel ...
+          & "C:\webrtc\build\tst\webrtc_client_test.exe" --gtest_filter="-SignalingApiFunctionalityTest.receivingIceConfigOffer_SlowClockSkew:SignalingApiFunctionalityTest.iceServerConfigRefreshConnectedAuthExpiration:SignalingApiFunctionalityTest.receivingIceConfigOffer_FastClockSkew:SignalingApiFunctionalityTest.receivingIceConfigOffer_FastClockSkew_VerifyOffsetRemovedWhenClockFixed:DataChannel ...
   # windows-msvc-mbedtls:
```
This is concerning so does h264 not work on Windows?  From the name of the test it looks like packing/unpacking the same frame results in a delta on Windows which is concerning, we should make a note to resolve this.

hassanctech commented on CMake/Dependencies/libkvsCommonLws-CMakeLists.txt:
```diff
@@ -4,21 +4,43 @@ project(libkvsCommonLws-download NONE)
+        -DUSE_MBEDTLS=${USE_MBEDTLS}
+        -DKVS_DEFAULT_STACK_SIZE=${KVS_DEFAULT_STACK_SIZE}
+        -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS}
+        -DPKG_CONFIG_EXECUTABLE=${PKG_CONFIG_EXECUTABLE}
```
```
-DPKG_CONFIG_EXECUTABLE=${PKG_CONFIG_EXECUTABLE}
```

Why is this necessary now but wasn't before?  I am not in favor of having this type of if/else it takes some effort to see what the difference between the two is.  We should find a way to consolidate this so we can either set a single variable that is different before the ExternalProject_Add, or export this variable PKG_CONFIG_EXECUTABLE and make a cmake global so it's available inside producer c build so we don't have to manually specify it.  This variable is set by when the system is trying to find package config.  This seems like instead there is a bug in the windows setup for producer c if it is not able to find package config on its own and needs it to be passed in, then we should fix it there if that is the case instead of duplicate this here.

niyatim23 replied:
I did this because we already do it for the Windows build in WebRTC. The error said it was specifically missing this path. I'm not sure what changed on Windows to require it now though. I'm using Divya's package manager, hoping to not do this now.

hassanctech commented on CMakeLists.txt:
```diff
@@ -254,7 +254,8 @@ set(BUILD_ARGS
         -DKVS_DEFAULT_STACK_SIZE=${KVS_DEFAULT_STACK_SIZE}
-        -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS})
+        -DCMAKE_C_FLAGS=${CMAKE_C_FLAGS}
+        -DPKG_CONFIG_EXECUTABLE=${PKG_CONFIG_EXECUTABLE})
```
We should not need to pass this, the build of the dependency should be searching for and finding pkg-config it shouldn't need to be told where it is.

disa6302 replied:
I fixed this in my current producer c PR to install pkgconfiglite for windows. So this should not be necessary.

niyatim23 replied:
Thanks, I've tried it out

niyatim23 replied:
Installing that package is not working for some reason: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/actions/runs/8741051317/job/23986196094?pr=1965#step:6:7565

hassanctech commented on src/source/PeerConnection/PeerConnection.c:
```diff
@@ -1563,6 +1563,10 @@ STATUS addTransceiver(PRtcPeerConnection pPeerConnection, PRtcMediaStreamTrack p
             depayFunc = depayVP8FromRtpPayload;
             clockRate = VIDEO_CLOCKRATE;
             break;
+        case RTC_CODEC_H265:
```
Do we need to specify a more descriptive name here that includes the profile / packetization mode like we have for h264?

niyatim23 replied:
That's because there is a default fmtp we use for h264 which is not the case for h265 as nothing has been enforced in the RFCs as far as I checked last. The offers that I received from Safari and Android do not contain fmtp / profile information either so I think it is more appropriate to use the current name

hassanctech replied:
ack

hassanctech commented on src/source/Rtp/Codecs/RtpH265Payloader.c:
```diff
@@ -0,0 +1,269 @@
+    CHK(nalus != NULL && pStart != NULL && pNaluLength != NULL, STATUS_NULL_ARG);
+
+    // Annex-B Nalu will have 0x000000001 or 0x000001 start code, at most 4 bytes
+    for (offset = 0; offset < 4 && offset < nalusLength && nalus[offset] == 0; offset++)
+        ;
```
Can we use a while loop here, with our current code stye we forbid a for loop with no { }.  I would suggest re-writing this something like:

Initialize offset to zero at the start of the function, then:

```
while(offset < 4 && offset < nalusLength && nalus[offset] == 0) {
    offset++;
}
```

niyatim23 replied:
This was copied over from the H264 payloader. The functions `getNextNaluLength` and `getNextNaluLengthH265` are identical. Would you recommend replacing it in both?

hassanctech replied:
Yes

niyatim23 replied:
Done!

hassanctech commented on src/source/Rtp/Codecs/RtpH265Payloader.c:
```diff
@@ -0,0 +1,269 @@
+    BOOL sizeCalculationOnly = (pPayloadArray == NULL);
+
+    CHK(nalu != NULL && filledLength != NULL && filledSubLenSize != NULL, STATUS_NULL_ARG);
+    sizeCalculationOnly = (pPayloadArray == NULL);
```
duplicate line, you have this already 3 lines above.

niyatim23 replied:
Removed from both, h264 and h265

niyatim23 commented on src/source/Rtp/Codecs/RtpH265Payloader.h:
```diff
@@ -0,0 +1,39 @@
+#endif
+
+#define FU_HEADER_SIZE 3
+#define FU_TYPE_ID     49
```
Done!

hassanctech reviewed (commented):
In the PR description it says it was tested with h265, I do not see any way to tell the master sample to use h265 it has hard coded h264 option.  Are those changes on a different branch somewhere else?  I do not see them in develop either.

niyatim23 commented:
> In the PR description it says it was tested with h265, I do not see any way to tell the master sample to use h265 it has hard coded h264 option. Are those changes on a different branch somewhere else? I do not see them in develop either.



> In the PR description it says it was tested with h265, I do not see any way to tell the master sample to use h265 it has hard coded h264 option. Are those changes on a different branch somewhere else? I do not see them in develop either.

I've added a command-line argument and updated the readme with new instructions. H264 and Opus are still the default.

disa6302 commented on README.md:
````diff
@@ -273,15 +273,18 @@ After executing `make` you will have sample applications in your `build/samples`
 
 ```shell
-./samples/kvsWebrtcClientMaster <channelName> 1
+./samples/kvsWebrtcClientMaster <channelName> 1 opus h265
````
Modify to: 
`./samples/kvsWebrtcClientMaster <channelName> 1 <audio-codec> <video-codec>` 

Allowed audio codec values: `aac, opus`
Allowed video codec values: `h264, h265`

niyatim23 replied:
Done

disa6302 commented on README.md:
````diff
@@ -292,23 +295,36 @@ Pass the desired media and source type when running the sample. The mediaType ca
 
+Using the testsrc with audio and video codec
+```shell
+./samples/kvsWebrtcClientMasterGstSample <channelName> audio-video testsrc opus h264
````
Let us generalize this. And add an example after the generalized command.

niyatim23 replied:
Done!

disa6302 commented on README.md:
````diff
@@ -292,23 +295,36 @@ Pass the desired media and source type when running the sample. The mediaType ca
+./samples/kvsWebrtcClientViewer <channelName> <audio-codec> <video-codec>
 ```
 
+Audio codec: aac, opus
````
Nit: `Allowed audio codec values`

disa6302 commented on samples/Common.c:
```diff
@@ -914,6 +913,8 @@ STATUS createSampleConfiguration(PCHAR channelName, SIGNALING_CHANNEL_ROLE_TYPE
     pSampleConfiguration->iceCandidatePairStatsTimerId = MAX_UINT32;
     pSampleConfiguration->pregenerateCertTimerId = MAX_UINT32;
     pSampleConfiguration->signalingClientMetrics.version = SIGNALING_CLIENT_METRICS_CURRENT_VERSION;
+    pSampleConfiguration->audioCodec = audioCodec;
```
Can we set these up in the samples directly instead of here after the `createSampleConfiguration` call? Only because we are changing the function signature.

niyatim23 replied:
Done!

disa6302 commented on samples/kvsWebRTCClientMaster.c:
```diff
@@ -25,7 +27,24 @@ INT32 main(INT32 argc, CHAR* argv[])
     pChannelName = argc > 1 ? argv[1] : SAMPLE_CHANNEL_NAME;
 #endif
 
-    CHK_STATUS(createSampleConfiguration(pChannelName, SIGNALING_CHANNEL_ROLE_TYPE_MASTER, TRUE, TRUE, logLevel, &pSampleConfiguration));
```
Would suggest keeping this here and then setting the pSampleConfiguration codec related params after we read the args.

niyatim23 replied:
Done

disa6302 commented on samples/kvsWebRTCClientMaster.c:
```diff
@@ -10,6 +10,8 @@ INT32 main(INT32 argc, CHAR* argv[])
     PCHAR pChannelName;
     SignalingClientMetrics signalingClientMetrics;
     signalingClientMetrics.version = SIGNALING_CLIENT_METRICS_CURRENT_VERSION;
+    RTC_CODEC audioCodec = RTC_CODEC_OPUS;
```
If these are the defaults, can we add a line in the readme to clarify that if the args are not provided, the default assumed by the sample is H264 and Opus?

disa6302 commented on samples/kvsWebRTCClientMasterGstSample.c:
```diff
@@ -3,6 +3,8 @@
 #include <gst/app/gstappsink.h>
 
 extern PSampleConfiguration gSampleConfiguration;
+RTC_CODEC audioCodec = DEFAULT_AUDIO_TRACK_ID;
```
Why do we need these if the values in PSampleConfiguration are being set and we can use those in `on_new_sample` callback?

niyatim23 replied:
Done

disa6302 commented on samples/kvsWebRTCClientMasterGstSample.c:
```diff
@@ -109,12 +111,12 @@ GstFlowReturn on_new_sample(GstElement* sink, gpointer data, UINT64 trackid)
 
 GstFlowReturn on_new_sample_video(GstElement* sink, gpointer data)
```
pSampleConfiguration is already passed as data here. We can use that to pass the appropriate track ID.

niyatim23 replied:
Done

niyatim23 replied:
I confused it with something else. Fixed. Thanks for pointing it out

disa6302 commented on src/source/Rtp/Codecs/RtpH264Payloader.c:
```diff
@@ -146,7 +147,6 @@ STATUS createPayloadFromNalu(UINT32 mtu, PBYTE nalu, UINT32 naluLength, PPayload
     BOOL sizeCalculationOnly = (pPayloadArray == NULL);
 
     CHK(nalu != NULL && filledLength != NULL && filledSubLenSize != NULL, STATUS_NULL_ARG);
-    sizeCalculationOnly = (pPayloadArray == NULL);
```
Why remove this?

niyatim23 replied:
Because we already check it on line 146

disa6302 commented on src/source/Rtp/Codecs/RtpH265Payloader.h:
```diff
@@ -0,0 +1,39 @@
+#define H265_FU_HEADER_SIZE 3
+#define H265_FU_TYPE_ID     49
+
+/*
```
Can we link the spec we referred to for this for future reference?

hassanctech commented on README.md:
````diff
@@ -273,15 +273,18 @@ After executing `make` you will have sample applications in your `build/samples`
 
 ```shell
-./samples/kvsWebrtcClientMaster <channelName> 1
+./samples/kvsWebrtcClientMaster <channelName> 1 <audio-codec> <video-codec>
````
When using Safari 17.4.1 with h265 enabled as JS viewer and C SDK as master with opus/h265 the video playback on JS side is constantly jumping back.

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/assets/7199741/55ff3d31-38d0-4fee-aa43-2860ce9180da

niyatim23 replied:
There was an issue with the source frames. Refer https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/1978

hassanctech commented on src/source/PeerConnection/SessionDescription.c:
```diff
@@ -420,12 +438,20 @@ STATUS populateSingleMediaSection(PKvsPeerConnection pKvsPeerConnection, PKvsRtp
         }
     }
+
+    if (RTC_CODEC_H265 == pRtcMediaStreamTrack->codec) {
+        payloadType = DEFAULT_PAYLOAD_H265;
+    }
```
Why do we need to do this? It looks like we do this again in LN 452-452.

niyatim23 replied:
Removed

hassanctech commented on src/source/PeerConnection/SessionDescription.c:
```diff
@@ -286,6 +301,9 @@ STATUS setTransceiverPayloadTypes(PHashTable codecTable, PHashTable rtxTable, PD
             (pKvsRtpTransceiver->transceiver.direction == RTC_RTP_TRANSCEIVER_DIRECTION_SENDRECV ||
              pKvsRtpTransceiver->transceiver.direction == RTC_RTP_TRANSCEIVER_DIRECTION_SENDONLY)) {
             CHK_STATUS(hashTableGet(codecTable, pKvsRtpTransceiver->sender.track.codec, &data));
+            if (pKvsRtpTransceiver->sender.track.codec == RTC_RTX_CODEC_H265) {
+                data = DEFAULT_PAYLOAD_H265;
+            }
```
This seems asymmetric compared to what we do for other codecs.  Why do we need to do this?

niyatim23 replied:
Removed. We don't.

hassanctech reviewed (commented):
A genera remark so different encoders (hardware/software) may have different quirks.  Aside from referencing the RFCs and following what should be supported, it is really important we look at a variety of encoders that are actually out in the wild and make sure this code is robust enough to work with them. At a minimum we should test with hardware and software encoders on iOS/Android and libx265 software encoder.  If this PR could include a matrix of encoders / combinations that were tested we can know in the future we are not regressing and we will know and have documented what gaps we have.

hassanctech reviewed (approved):
I haven't looked at the RFC myself to finely go through the payloader/depayloader code will assume the proper due diligence was done there.

disa6302 reviewed (approved):
LGTM.

niyatim23 commented:
> A general remark so different encoders (hardware/software) may have different quirks. Aside from referencing the RFCs and following what should be supported, it is really important we look at a variety of encoders that are actually out in the wild and make sure this code is robust enough to work with them. At a minimum we should test with hardware and software encoders on iOS/Android and libx265 software encoder. If this PR could include a matrix of encoders / combinations that were tested we can know in the future we are not regressing and we will know and have documented what gaps we have.

Tested the following combinations
| Master | Viewer |
|-----------------|-----------------|
| kvsWebrtcClientMaster    | Safari Version 17.4.1    |
| kvsWebrtcClientMaster    | Android phone (Samsung SM-S908U Android 12) with [Sample](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-android) with H265 libs [here](https://github.com/hongliwo/amazon-kinesis-video-streams-webrtc-sdk-android/tree/3f2424d029ec095de1b0322fc27b72cb6a76dcc8/libs)    |
| kvsWebrtcClientMaster   | kvsWebrtcClientViewerGstSample (ffplay with file obtained from filesink)    |
| kvsWebrtcClientMasterGstSample    | Safari Version 17.4.1    |
| kvsWebrtcClientMasterGstSample    | Android phone (Samsung SM-S908U Android 12) with [Sample](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-android) with H265 libs [here](https://github.com/hongliwo/amazon-kinesis-video-streams-webrtc-sdk-android/tree/3f2424d029ec095de1b0322fc27b72cb6a76dcc8/libs)    |
| kvsWebrtcClientMasterGstSample   | kvsWebrtcClientViewerGstSample (ffplay with file obtained from filesink)   |

naseebpanghal commented:
Hi @niyatim23,
Is it possible to run kvsWebrtcClientMaster	kvsWebrtcClientViewerGstSample  both on same machine locally?

niyatim23 commented:
>Is it possible to run kvsWebrtcClientMaster kvsWebrtcClientViewerGstSample both on same machine locally?

@naseebpanghal, yes it should be

naseebpanghal commented:
Hi @niyatim23 
I am trying but not able to stream. I doing all this to test H265 stream which is not supported by various web browers so far. That is the reason I have chosen kvsWebrtcClientViewerGstSample as client(mentioned by you above in this post).

[client.log](https://github.com/user-attachments/files/16398720/client.log)
[server.log](https://github.com/user-attachments/files/16398721/server.log)

First i tried with H264, it works with web browser but not with local kvsWebrtcClientViewerGstSample running on same machine.
